frame/staking/src/pallet/mod.rs

```diff
@@ -559,6 +651,184 @@ pub mod pallet {
+		pub(crate) fn get_page_count(era: EraIndex, validator: &T::AccountId) -> PageIndex {
+			<ErasStakersOverview<T>>::get(&era, validator).page_count.max(1)
+		}
+
+		/// Returns the next page that can be claimed or `None` if nothing to claim.
+		pub(crate) fn get_next_claimable_page(
+			era: EraIndex,
+			validator: &T::AccountId,
+			ledger: &StakingLedger<T>,
+		) -> Option<PageIndex> {
+			if Self::is_non_paged_exposure(era, validator) {
+				return match ledger.legacy_claimed_rewards.binary_search(&era) {
+					// already claimed
+					Ok(_) => None,
+					// Non-paged exposure is considered as a single page
+					Err(_) => Some(0),
+				}
+			}
+
+			// Find next claimable page of paged exposure.
+			let page_count = Self::get_page_count(era, validator);
+			let claimed_pages = ClaimedRewards::<T>::get(era, validator);
+			let claimed_page_count = claimed_pages.len() as PageIndex;
+
+			// find the first page that is not claimed.
+			for page in 0..claimed_page_count as PageIndex {
+				debug_assert!(page <= claimed_pages[page as usize]);
+				if page < claimed_pages[page as usize] {
+					return Some(page)
+				}
+			}
+			// all pages are claimed
+			return if claimed_page_count < page_count { Some(claimed_page_count) } else { None }
+		}
+
+		/// Checks if exposure is paged or not.
+		fn is_non_paged_exposure(era: EraIndex, validator: &T::AccountId) -> bool {
+			<ErasStakersClipped<T>>::contains_key(&era, validator)
+		}
+
+		/// Returns the maximum number of pages of exposure we can store.
+		fn get_max_exposure_page_count() -> PageIndex {
+			return <MaxExposurePageCount<T>>::get().unwrap_or_default().max(1)
+		}
+
+		/// Returns validator commission for this era and page.
+		pub(crate) fn get_validator_commission(
+			era: EraIndex,
+			validator_stash: &T::AccountId,
+		) -> Perbill {
+			<ErasValidatorPrefs<T>>::get(&era, validator_stash).commission
+		}
+
+		/// Creates an entry to track validator reward has been claimed for a given era and page.
+		pub(crate) fn set_rewards_as_claimed(
+			era: EraIndex,
+			validator: &T::AccountId,
+			page: PageIndex,
+		) {
+			let mut claimed_pages = ClaimedRewards::<T>::get(era, validator);
+			let search = claimed_pages.binary_search(&page);
+			// this should never be called if the reward has already been claimed
+			debug_assert!(search.is_err());
+
+			match search {
+				Err(index) => {
+					claimed_pages.insert(index, page);
+					ClaimedRewards::<T>::insert(era, validator, claimed_pages);
+				},
+				_ => {},
+			}
+		}
+
+		/// Store exposure for elected validators at start of an era.
+		pub(crate) fn set_validator_exposure(
+			era: EraIndex,
+			validator: &T::AccountId,
+			exposure: Exposure<T::AccountId, BalanceOf<T>>,
+		) {
+			<ErasStakers<T>>::insert(era, &validator, &exposure);
+
+			let page_size = <ExposurePageSize<T>>::get()
+				.unwrap_or_else(|| T::MaxNominatorRewardedPerValidator::get())
+				.clamp(1, T::MaxNominatorRewardedPerValidator::get());
```

Comment:
I think the `clamping` is fine, but some notes: 

1. you probably only want to do `.max(1)` right? 
2. It is technically a `defensive_clamp` 💡, if all assumptions are held? if that's the case, you can make a separate PR for `defensive_clamp`. We already have `defensive_max` etc IIRC.

Reply:
`defensive_clamp` makes sense. I do want to clamp since the size should not exceed `MaxNominatorRewardedPerValidator` (or actually it would be `MaxExposurePageSize`) as well. But both of this should be taken care of in the setter for `ExposurePageSize` and should be checked in `integrity_test` as well.